File: Programas/NoInformados/Rompecabezas/2. Busqueda en Amplitud/JavaScript BFS/Rompecabezas.js
import * as Gestor from './Gestor.js'

export class Rompecabezas{

    constructor( nodoPadre=null , accion=null , profundidad=0 ){
        this.nodoPadre = nodoPadre
        this.profundidad = profundidad
        if( !nodoPadre ){
            this.celdas = Gestor.secuencia.reduce( ( res , sig ) => res | ( BigInt(sig) << BigInt(4*(sig-1)) ) , 0n )
            this.cero = 8
        }
        else{
            this.celdas = nodoPadre.celdas
            this.cero = nodoPadre.cero
            if(accion)
                this.mover(accion)
        }
    }

    mostrarRompecabezas(){
        let tabla = ""
        for( let i = 0 ; i < 9 ; i++ ){
            tabla += Gestor.obtenerCelda( i , this.celdas ) + " "
            if( i % 3 == 2 )
                tabla += "\n" 
        }
        return tabla
    }

    mover( accion ){
        if(accion =='N')
            Gestor.mover_celda_arriba(this)
        if(accion =='E')
            Gestor.mover_celda_derecha(this)
        if(accion =='S')
            Gestor.mover_celda_abajo(this)
        if(accion =='O')
            Gestor.mover_celda_izquierda(this)
    }

    compararCeldas( otro ){
        return otro.celdas == this.celdas
    }

    expandir(){
        let nodosHijos = []
        for( let movimiento of Gestor.movimientos_validos[ this.cero ]  ){
            let nodo = new Rompecabezas( this , movimiento , this.profundidad+1 )
            if( this.nodoPadre ){
                if( nodo.celdas != this.nodoPadre.celdas )
                    nodosHijos.push( nodo )
            }else{
                nodosHijos.push( nodo )
            }
        }
        return nodosHijos
    }

    modificarCeldas( tablero ){
        this.celdas = 0n
        for (let i = tablero.length-1 ; i >= 0 ; i--) {
            this.celdas = this.celdas << 4n  | BigInt(tablero[i])
            if( tablero[i] == 0 )
                this.cero = i
        }
    }

}

File: Programas/NoInformados/Rompecabezas/1. Busqueda en Profundidad v1.0.0/PYTHON DFS/Rompecabezas.py
from functools import reduce
import random
import Gestor

class Rompecabezas:

    # Constructor
    def __init__(self, nodoPadre=None, accion=None, profundidad=0):

        self.nodoPadre = nodoPadre
        self.profundidad = profundidad

        if( nodoPadre == None ):
            self.tablero = reduce( lambda resultado , elemento: resultado | ( elemento << 4*(elemento-1) ), Gestor.elementos )
            self.cero = 8
        else:
            self.tablero = nodoPadre.tablero
            self.cero = nodoPadre.cero
            if( accion != None):
                self.mover( accion )

    # Metodo para mostrar el tablero 
    def __str__(self):
        tabla = ""
        for i in Gestor.elementos:
            tabla = tabla + str(Gestor.obtenerCelda( i , self.tablero )) + " "
            if (i % 3 == 2):
                tabla = tabla + "\n"
        return tabla

    # Metodo para comparar los tableros
    def compararCeldas(self, otro):
        return otro.tablero == self.tablero

    # Metodo para mover una celda
    def mover(self, accion):
        if accion == 'N':
            Gestor.mover_celda_arriba(self)
        elif accion == 'E':
            Gestor.mover_celda_derecha(self)
        elif accion == 'S':
            Gestor.mover_celda_abajo(self)
        elif accion == 'O':
            Gestor.mover_celda_izquierda(self)

    ## Metodo para revolver el rompecabezas
    def revolverCeldas(self, numeroMovimientos):
        for i in range(0,numeroMovimientos):
            self.mover(random.choice(Gestor.movimientos_validos[self.cero]))

    # Metodo para expandir el nodo padre
    def expandirNodos(self):
        nodos = list()
        for movimiento in Gestor.movimientos_validos[ self.cero ]:
            nodo = Rompecabezas( self , movimiento , self.profundidad+1 )
            if(self.nodoPadre != None):
                if(nodo.tablero != self.nodoPadre.tablero):
                    nodos.append(nodo)
            else:
                nodos.append(nodo)
        return nodos

